File: PJT/PJT_10/proxy-server.js
/**
 * proxy-server.js
 * Kakao Mobility API CORS 우회용 로컬 프록시 서버
 *
 * [실행 방법]
 *   node proxy-server.js
 *
 * [주의] Node.js 내장 모듈만 사용 - npm install 불필요!
 *
 * 서버 포트: 3001
 * 브라우저에서 http://localhost:3001/directions?... 으로 요청하면
 * 이 서버가 Kakao Mobility API에 인증 헤더와 함께 실제 요청을 보내고
 * 결과를 브라우저에 그대로 돌려줍니다.
 */

const http  = require('http');
const https = require('https');
const url   = require('url');

// apikey.js에서 직접 키를 가져올 수 없으므로 여기에 직접 입력
const REST_API_KEY = 'a491c4093faf6ff1903948840b8abe1f';
const PORT = 3001;

const server = http.createServer((req, res) => {

  // ── CORS 허용 헤더 (Live Server http://127.0.0.1:5500 허용) ──
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  // Preflight 요청(OPTIONS) 처리
  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  const parsedUrl = url.parse(req.url, true);
  const pathname  = parsedUrl.pathname;

  // ── /directions 경로만 프록시 처리 ──
  if (req.method === 'GET' && pathname === '/directions') {
    const query = parsedUrl.search || '';   // ?origin=...&destination=... 부분

    const options = {
      hostname: 'apis-navi.kakaomobility.com',
      path: `/v1/directions${query}`,
      method: 'GET',
      headers: {
        'Authorization': `KakaoAK ${REST_API_KEY}`,
        'Content-Type' : 'application/json'
      }
    };

    console.log(`[프록시] 요청 전달: ${options.hostname}${options.path}`);

    const proxyReq = https.request(options, (proxyRes) => {
      let body = '';
      proxyRes.setEncoding('utf8');
      proxyRes.on('data', chunk => { body += chunk; });
      proxyRes.on('end', () => {
        console.log(`[프록시] 응답 수신 - 상태: ${proxyRes.statusCode}`);
        res.writeHead(proxyRes.statusCode, { 'Content-Type': 'application/json; charset=utf-8' });
        res.end(body);
      });
    });

    proxyReq.on('error', (err) => {
      console.error('[프록시] 요청 오류:', err.message);
      res.writeHead(500, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: err.message }));
    });

    proxyReq.end();

  } else {
    // 다른 경로는 404
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: '지원하지 않는 경로입니다.' }));
  }
});

server.listen(PORT, '127.0.0.1', () => {
  console.log('');
  console.log('========================================');
  console.log('  🚗 카카오 모빌리티 프록시 서버 실행 중');
  console.log(`  URL: http://127.0.0.1:${PORT}`);
  console.log('========================================');
  console.log('  Live Server와 함께 이 창을 열어두세요.');
  console.log('  종료하려면 Ctrl+C 를 누르세요.');
  console.log('========================================');
  console.log('');
});

File: PJT/PJT_10/proxy_server.py
"""
proxy_server.py
Kakao Mobility API CORS 우회용 로컬 프록시 서버

[실행 방법]
  python proxy_server.py

[주의] Python 3 내장 모듈만 사용 - pip install 불필요!

서버 포트: 3001
브라우저에서 http://localhost:3001/directions?... 로 요청하면
이 서버가 Kakao Mobility API에 인증 헤더와 함께 실제 요청을 보내고
결과를 브라우저에 그대로 돌려줍니다.
"""

import http.server
import urllib.request
import urllib.parse
import json
from urllib.error import HTTPError, URLError

# ──────────────────────────────────────────
REST_API_KEY = "a491c4093faf6ff1903948840b8abe1f"
PORT = 3001
# ──────────────────────────────────────────


class ProxyHandler(http.server.BaseHTTPRequestHandler):

    def log_message(self, format, *args):
        """로그 포맷 커스텀"""
        print(f"[프록시] {self.address_string()} - {format % args}")

    def _send_cors_headers(self, status=200, content_type="application/json; charset=utf-8"):
        self.send_response(status)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.send_header("Content-Type", content_type)
        self.end_headers()

    def do_OPTIONS(self):
        """Preflight 요청 처리"""
        self._send_cors_headers(204)

    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)

        if parsed.path != "/directions":
            self._send_cors_headers(404)
            self.wfile.write(json.dumps({"error": "지원하지 않는 경로입니다."}).encode("utf-8"))
            return

        # Kakao Mobility API로 프록시
        query = parsed.query
        target_url = f"https://apis-navi.kakaomobility.com/v1/directions?{query}"
        print(f"[프록시] → {target_url}")

        req = urllib.request.Request(
            target_url,
            headers={
                "Authorization": f"KakaoAK {REST_API_KEY}",
                "Content-Type": "application/json",
            },
            method="GET",
        )

        try:
            with urllib.request.urlopen(req) as response:
                body = response.read()
                status = response.status
                print(f"[프록시] ← 응답 상태: {status}")
                self._send_cors_headers(status)
                self.wfile.write(body)

        except HTTPError as e:
            body = e.read()
            print(f"[프록시] ← HTTP 에러: {e.code}")
            self._send_cors_headers(e.code)
            self.wfile.write(body)

        except URLError as e:
            print(f"[프록시] ← 연결 오류: {e.reason}")
            self._send_cors_headers(500)
            self.wfile.write(json.dumps({"error": str(e.reason)}).encode("utf-8"))


if __name__ == "__main__":
    server = http.server.HTTPServer(("127.0.0.1", PORT), ProxyHandler)
    print()
    print("=" * 45)
    print("  [Kakao Mobility Proxy Server]")
    print(f"  URL: http://127.0.0.1:{PORT}")
    print("=" * 45)
    print("  Live Server와 함께 이 창을 열어두세요.")
    print("  종료하려면 Ctrl+C 를 누르세요.")
    print("=" * 45)
    print()
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n[프록시] 서버를 종료합니다.")
        server.shutdown()
